### DISCUSSION TITLE
Danger: please do NOT recommend that Docker upgrade guide!

### DISCUSSION THREAD

**Discussion by Kixunil** [btcpayserver/btcpayserver#3371] [General]
The upgrade guide recommends doing this super-dangerous step: `sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 04EE7237B7D453EC 648ACFD622F3D138`

This has two problems:

* These are **short-version** fingerprints which means it may be possible to find a collision and attack the users
* The users have no good way to check the fingerprints, I checked the fingerprints and they are "OK" (still collision risk) but people shouldn't do this unless they really, really checked **and entered the full** fingerprints.

`man apt-key` also warns:

```
adv
    Pass advanced options to gpg. With adv --recv-key you can e.g. download key from keyservers directly into the trusted set of keys. Note that there are no checks performed, so it is easy to
    completely undermine the apt-secure(8) infrastructure if used without care.
```

Note that from my understanding of it the download also isn't sandboxed so a hypothetical vulnerability in GPG *download* code could cause root-level compromise. Not very likely but grave consequences.

Thankfully the author is at least smart enough to tell people about the priority thing. For the future, I offer you "free" security reviews for Linux-related things like this. ("Free" in quotes because you earned it by creating and maintaining BTCPay :))

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA1

I hereby declare that I observed these two keys being installed on my Debian 10 Qubes VM template by default:

E1CF 20DD FFE4 B89E 8026  58F1 E0B1 1894 F66A EC98
80D1 5823 B7FD 1561 F9F7  BCDD DC30 D7C2 3CBB ABEE

I did not perform any action that could compromise them, so assuming Qubes developers were not compromised and Qubes security holds these should be valid keys of Debian 9 and 10 repositories.
-----BEGIN PGP SIGNATURE-----

iQIcBAEBAgAGBQJh8vshAAoJEOxMOYS+q5a/DXwP/0Jgxn1okp4oWpAw104aEJ2o
fdHbHyT1SGhJ8n1Lou+n6mBhlWxhbxtL7OFWPZ3FgOp4dwlQoP1JGUtm66jHLpyO
kNU8CaS1OBgu49XPsx9OlrU8E005e44mJ5JWZosLriD+MlSe12nhkrHNjwEJBmOb
qVrooitXXXGsSdccl6qvuHUwHiTtkWNjL4M+JWbbcR8+RWBIR1imtPpdS2lpRi0r
pgdIy4rIG38Yb6IT+ZBps5V9igllR+hZYtmMYPUQy8ec/SBLvVfuEqkXJRb/iAIy
CH/t5c/N/rzAL2SGn5YfYd7w1M2OmBjRZ96QBlY6fM0UXTbny4dU+cuNJCLEofoj
0jSPvmqi/yjy9p/sv7+IeeemciAdFgF1Wxkm1nuU969KNPdgnh9B0vM/OUNIozUt
MfR7Kob4Lv7rbwdGMVW8kKaXUdANy/hsZ75LdXdPlYwaav4E/4hwFKQaHTQKlIFk
313qkCpIxEMvg4rPBlVNDERHA/s+RmXAaKuNQiBKUR5lAE3iGJqqZbwsaBqp3BW2
SEGZryx0cLeH9TrTaks1kNG1eg3lXpiKkm8934xNs1tpez1hr1nPyeAC1bd204An
GrbZwxe5szT37PtIJUyZNVfImtN4b6JseLWR4/IkoM61fzx3UFvhwQvusWYAUB+d
5d16hDU6QzuhYfwYL9Ch
=D5bm
-----END PGP SIGNATURE-----
```

**Comment by Kixunil** [comment]
Speaking of security, are the login codes single-use?

**Reply by mp21-org** [reply]
Could you please link to the "upgrade guide" you are refering to? I've searched on our https://docs.btcpayserver.org/ page and the btcpayserver-docker repo for the short GPG keys you mentioned, and hadn't found them.

It seems like an issue worth escalating, if it is still an issue.

sorry that has been apparently neglected for way too long.

**Reply by Kixunil** [reply]
Hmm, I don't remember anymore where the link was, it pointed at an external site. But when trying to find it I discovered that [some script downsloads a key without verifying it](https://github.com/btcpayserver/btcpayserver-doc/blob/90121d220ba88bbf03bb8007a494fc1157e5d28a/docs/Deployment/btcpayserver-rpi4-install.md#L16)

Auto [gpg verifying can be tricky](https://github.com/lightningnetwork/lnd/pull/5048), I recommend downloading into a temporary keyring and then exporting the key with specified full fingerprint. That way multiple keys can't be added by an attacker.

**Reply by mp21-org** [reply]
Please report further security related findings directly via email, before public disclosure:
https://docs.btcpayserver.org/BTCPayServer/Security/

Be assured that these kinds of findings are very valuable to the project.

**Reply by NicolasDorier** [reply]
It is following the official way of installing docker https://docs.docker.com/engine/install/ubuntu/
